ProcessLib/ThermoHydroMechanics/ThermoHydroMechanicsFEM.h

```diff
@@ -0,0 +1,889 @@
+        local_Jac
+            .template block<pressure_size, temperature_size>(pressure_index,
+                                                             temperature_index)
+            .noalias() -= storage_T / dt;
+
+        // pressure equation, pressure part.
+        local_Jac
+            .template block<pressure_size, pressure_size>(pressure_index,
+                                                          pressure_index)
+            .noalias() += laplace_p + storage_p / dt;
+
+        // pressure equation, displacement part.
+        local_Jac
+            .template block<pressure_size, displacement_size>(
+                pressure_index, displacement_index)
+            .noalias() += Kup.transpose() / dt;
+
+        // pressure equation (f_p)
+        local_rhs.template segment<pressure_size>(pressure_index).noalias() -=
+            laplace_p * p + storage_p * p_dot - storage_T * T_dot +
+            Kup.transpose() * u_dot;
+
+        // displacement equation (f_u)
+        local_rhs.template segment<displacement_size>(displacement_index)
+            .noalias() += Kup * p;
+
+        // temperature equation (f_T)
+        local_rhs.template segment<temperature_size>(temperature_index)
+            .noalias() -= KTT * T + MTT * T_dot;
+    }
+
+    void postTimestepConcrete(std::vector<double> const& local_x) override
+    {
+        double const& t = _process_data.t;
+
+        auto p =
+            Eigen::Map<typename ShapeMatricesTypePressure::template VectorType<
+                pressure_size> const>(local_x.data() + pressure_index,
+                                      pressure_size);
+
+        auto T =
+            Eigen::Map<typename ShapeMatricesTypePressure::template VectorType<
+                pressure_size> const>(local_x.data() + temperature_index,
+                                      temperature_size);
+
+        using GlobalDimVectorType =
+            typename ShapeMatricesTypePressure::GlobalDimVectorType;
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        SpatialPosition x_position;
+        x_position.setElementID(_element.getID());
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            auto const& N_T = _ip_data[ip].N_p;
+            auto const& N_p = N_T;
+            auto T_int_pt = N_T * T;
+            auto p_int_pt = N_p * p;
+
+            x_position.setIntegrationPoint(ip);
+            double const K_over_mu =
+                _process_data.intrinsic_permeability(t, x_position)[0] /
+                _process_data.getFluidViscosity(p_int_pt, T_int_pt);
+
+            auto const rho_f =
+                _process_data.getFluidDensity(p_int_pt, T_int_pt);
+            auto const& b = _process_data.specific_body_force;
+
+            // Compute the velocity
+            auto const& dNdx_p = _ip_data[ip].dNdx_p;
+            GlobalDimVectorType const darcy_velocity =
```

Comment:
The calculation of velocity is going to be changed. Maybe this change can be applied here in another PR.

Reply:
Here the same. Please add if (!b.empty()) condition,

Reply:
@wenqing `b` has type `Eigen::Matrix<double, DisplDim, 1>` which has no `empty()` member.

Reply:
@endJunction I see that b is always initialized. Sorry I am wrong here.